linked-list.rb

```diff
@@ -3,73 +3,101 @@
-       if current.value == val
-           # If the head is the element to be delete, the head needs to be updated
-           @head = @head.next_node
-       else
-           # ... x -> y -> z
-           # Suppose y is the value to be deleted, you need to reshape the above list to :
-           #   ... x->z
-           # ( and z is basically y.next_node )
-           current = @head
-           while (current != nil) && (current.next_node != nil) && ((current.next_node).value != val)
-               current = current.next_node
-           end
-
-           if (current != nil) && (current.next_node != nil)
-               current.next_node = (current.next_node).next_node
-           end
-       end
-   end
-
-   def display
-       # Traverse through the list till you hit the "nil" at the end
-       current = @head
-       full_list = []
-       while current.next_node != nil
-           full_list += [current.value.to_s]
-           current = current.next_node
-       end
-       full_list += [current.value.to_s]
-       puts full_list.join("->")
-   end
-
-   def include?(key)
-   end
-
-   def size
-   end
-
-   def max
-   end
+  def initialize(val)
+    # Initialize a new node at the head
+    @head = Node.new(val,nil)
+  end
+
+  def add(value)
+    # Traverse to the end of the list
+    # And insert a new node over there with the specified value
+    current = @head
+    while current.next_node != nil
+      current = current.next_node
+    end
+    current.next_node = Node.new(value,nil)
+    self
+  end
+
+  def delete(val)
+    current = @head
+    if current.value == val
+      # If the head is the element to be delete, the head needs to be updated
+      @head = @head.next_node
+    else
+      # ... x -> y -> z
+      # Suppose y is the value to be deleted, you need to reshape the above list to :
+      #   ... x->z
+      # ( and z is basically y.next_node )
+      current = @head
+      while (current != nil) && (current.next_node != nil) && ((current.next_node).value != val)
+        current = current.next_node
+      end
+
+      if (current != nil) && (current.next_node != nil)
+        current.next_node = (current.next_node).next_node
+      end
+    end
+  end
+
+  def display
+    # Traverse through the list till you hit the "nil" at the end
+    current = @head
+    full_list = []
+    while current.next_node != nil
+      full_list += [current.value.to_s]
+      current = current.next_node
+    end
+    full_list += [current.value.to_s]
+    puts full_list.join("->")
+  end
+
+  # HOMEWORK: next 3 methods
+  def include?(key)
+    current = @head
+    while current.next_node != nil
+      return true if current.value == key
+      current = current.next_node
+    end
+
+    return true if current.value == key
+
+    false
+  end
+
+  def size
+    current = @head
+    list_size = 0
+    while current.next_node != nil
+      list_size += 1
+      current = current.next_node
+    end
+
+    list_size += 1
+  end
```

Comment:
This looks great and works, but in all of these (include?, size, and max) you had to account for that last node separately -- if you change your loop to `while current != nil` instead of `while current.next_node != nil` then you wouldn't need that! I told you to model after display which does it the latter way -- because sometimes you need to do something different with the last case and I wanted that to be the pattern I showed -- but now I want to point out how we could have optimized the pattern just slightly. Hopefully that makes sense!